Make possible to favorite properties in the inspector

YeldhamDev opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/4274.

https://github.com/user-attachments/assets/9b331d7b-e99b-43ea-b5d4-16e109ece6a8

## Limitations

Incompatible with built-in scripts, as validating the favorites file with them in mind would mean extra loading of scenes at startup.

**Sponsored By:** 🐺 Lone Wolf Technology / 🍀 W4 Games.

YeldhamDev committed:
Make possible to favorite properties in the inspector

timothyqiu commented:
Favoriting `Button`'s `tooltip_text` and `text` makes two "Text" properties :rofl:

![image](https://github.com/user-attachments/assets/4671ffb8-1458-4eb6-af48-89c829c89ab1)

The group "Tooltip" is lost here.

If this is not expected, I think displaying "Tooltip Text" is not an ideal solution as it means doubling translator's work for grouped properties. Maybe displaying a group that's always unfolded?

Mickeon commented:
This is not necessarily related to this proposal as the rest of the Inspector suffers from it too, but I don't like that there's no way to know that Inspector categories can be right-clicked (not even a text color on hover). This makes the "Unfavourite all" option undiscoverable.

tetrapod00 commented:
Could we make `Node3D.Position`, `Node3D.Rotation`, and `Node3D.Scale` (and the corresponding 2D properties) favorites by default? This would make the default editor show those properties at the top of the inspector, which is a very common request. It would also increase the discoverability of this feature.

ryevdokimov commented:
> Could we make `Node3D.Position`, `Node3D.Rotation`, and `Node3D.Scale` (and the corresponding 2D properties) favorites by default? This would make the default editor show those properties at the top of the inspector, which is a very common request. It would also increase the discoverability of this feature.

I think the inspector can get crowded pretty easily and would prefer not to have additional and redundant data shown by default.

tetrapod00 commented:
> I think the inspector can get crowded pretty easily and would prefer not to have additional and redundant data shown by default.

Notably this implementation pulls properties to the top, it doesn't duplicate them. Setting these properties as favorites by default *does* obscure the fact that they are Node3D properties, though. 
![godot windows editor x86_64_sPe583lPty](https://github.com/user-attachments/assets/ac7a4df8-6bbe-4cf8-8dfb-09546f78a6a7)
![godot windows editor x86_64_QlHUVfnqBK](https://github.com/user-attachments/assets/6072263c-cb53-42ca-b46f-6d51058e9513)

ryevdokimov commented:
> Notably this implementation pulls properties to the top, it doesn't duplicate them. Setting these properties as favorites by default does obscure the fact that they are Node3D properties, though.

I should have paid attention more, thanks for pointing that out. That is surprising though, and not what I would expect for the reason you stated.

jaydensipe commented:
> Setting these properties as favorites by default _does_ obscure the fact that they are Node3D properties, though.

Wondering if something could be done like this, where it would show the Object/Class name before the property?

![test](https://github.com/user-attachments/assets/7889c2e3-067b-4b88-81c7-9722d53d94a7)

tetrapod00 commented:
My view is that for most workflows involving the editor, you want to access position, rotation, and scale most often, at the top of the inspector, then you want to access properties you defined on your classes, then you want to access properties on built-in parent classes. As is, the ordering of properties is mostly good except for the fact that transform properties are hidden at the bottom of the list.

If *most people* setting up their editor will favorite these, then we should do it by default. The favorites are stored in editor settings and are per-project, so whichever default is chosen does matter - each new project will be affected by it. It will be three clicks to change it, whichever default we choose, so let's choose the default that serves the most people. Of course, if the preference is evenly split, we can decide in favor of no default favorites.

ryevdokimov commented:
>If most people setting up their editor will favorite these, then we should do it by default.

Agreed.

If/when this gets approved and merged, I would suggest a separate proposal and PR be made.

tetrapod00 commented:
Also, to bikeshed the name, I do slightly prefer the old "Pinned Properties" name from the original proposal.

Mickeon commented:
If we settle on "pinning" properties, the "Pin property" feature has to be renamed to something else. To be very fair, its name has been potentially misleading as is, already.

tetrapod00 commented:
> If we settle on "pinning" properties, the "Pin property" feature has to be renamed to something else. To be very fair, its name has been potentially misleading as is, already.

Oh, I *thought* that there was already something called "property pinning" but googling that led me to the original proposal for this new feature. "Pinning properties" definitely describes this feature more than the existing one. But changing the name of an existing feature is probably too much, just to have a slightly better name for the new feature.

KoBeWi commented:
It can be called Sticky Properties or something similar.

YeldhamDev commented:
Alright, now properties are shown with their respective sections (always unfolded), to both give context, and prevent ones with the same name mixing up.
![image](https://github.com/user-attachments/assets/48d7d467-68c1-4836-8289-cd0bf426c4e7)

pineapplemachine commented:
I foresee not finding properties in the expected place, because they are moved rather than duplicated, being a potential source of confusion and irritation for larger projects where you might not remember what properties were favorited and might have scrolled past the favorites category without noticing the property there.

At minimum I'd really like to see an editor toggle for whether favorited properties are moved or duplicated.

KoBeWi commented:
Favorite and Pin should be in the same separator group.
![image](https://github.com/user-attachments/assets/741494d8-431a-4919-81ca-d3bb241ed95d)

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1695,21 +1695,33 @@ String FileSystemDock::_get_unique_name(const FileOrFolder &p_entry, const Strin
-	EditorSettings::get_singleton()->set_favorites(new_favorites);
+		String old_path = I->key;
+		if (p_files_renames.has(old_path)) {
+			favorite_properties.replace_key(old_path, p_files_renames[old_path]);
```
Wouldn't it be simpler to instead iterate on renames and replace favorites if they match?

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1695,21 +1695,33 @@ String FileSystemDock::_get_unique_name(const FileOrFolder &p_entry, const Strin
+		++next;
 
-	EditorSettings::get_singleton()->set_favorites(new_favorites);
+		String old_path = I->key;
```
```suggestion
		const String &old_path = I->key;
```

KoBeWi commented on editor/editor_settings.h:
```diff
@@ -102,6 +102,7 @@ class EditorSettings : public Resource {
 	HashMap<String, List<Ref<InputEvent>>> builtin_action_overrides;
 
 	Vector<String> favorites;
+	HashMap<StringName, PackedStringArray> favorite_properties;
```
I wonder if it shouldn't be String instead of StringName. You seem to be doing some String conversions, and also some keys can be script paths.

KoBeWi commented on editor/editor_settings.cpp:
```diff
@@ -1497,10 +1497,35 @@ void EditorSettings::set_favorites(const Vector<String> &p_favorites) {
+	favorite_properties = p_favorite_properties;
+	String favorite_properties_file;
+	if (Engine::get_singleton()->is_project_manager_hint()) {
+		favorite_properties_file = EditorPaths::get_singleton()->get_config_dir().path_join("favorite_properties");
```
When is this relevant for Project Manager?

YeldhamDev replied:
Copy-and-paste mistake, fixed.

KoBeWi commented on editor/editor_settings.cpp:
```diff
@@ -1497,10 +1497,35 @@ void EditorSettings::set_favorites(const Vector<String> &p_favorites) {
+
+	Ref<ConfigFile> cf;
+	cf.instantiate();
+	for (HashMap<StringName, PackedStringArray>::ConstIterator I = p_favorite_properties.begin(); I;) {
```
Why not use KeyValue range iteration?

YeldhamDev replied:
Been a little while since I used those iterators, so the reason is: i forgor :skull:

KoBeWi replied:
You used them in #97257 though 🤔
Also remember about KV.

YeldhamDev replied:
> You used them in https://github.com/godotengine/godot/pull/97257 though 🤔

My goldfish memory combined with how busy my week has been makes anything past a day or two feel like a millennia.

> Also remember about KV.

Some questions about the `KV` stuff:
- Is it `KV` or `kv`? `kv` seems to be used the most.
- What's the convention when there's a `KeyValue` iteration inside another?

KoBeWi replied:
- KV. It applies to new code, there are no plans to "fix" the already existing loops.
- No idea. IMO "KV2" would be fine, but maybe ask on the chat, or just use whatever.

KoBeWi commented on editor/editor_settings.cpp:
```diff
@@ -1523,23 +1548,59 @@ Vector<String> EditorSettings::get_recent_dirs() const {
+						favorite_properties[E].push_back(properties[i]);
+					}
+				}
+			} else {
+				if (ResourceLoader::exists(E, "Script")) {
```
```suggestion
			} else if (ResourceLoader::exists(E, "Script")) {
```

KoBeWi commented on editor/editor_settings.cpp:
```diff
@@ -1523,23 +1548,59 @@ Vector<String> EditorSettings::get_recent_dirs() const {
+			PackedStringArray properties = PackedStringArray(cf->get_value(E, "properties"));
+			if (EditorNode::get_editor_data().is_type_recognized(E)) {
+				for (int i = 0; i < properties.size(); i++) {
+					if (ClassDB::has_property(E, properties[i], true), !favorite_properties.has(properties[i])) {
```
What's the first check for? Native types won't just randomly lose properties, it's relevant for Scripts if anything.

YeldhamDev replied:
This is basically just to check if whatever comes from the saved file is valid.

KoBeWi replied:
If the property doesn't exist, most likely the file was modified manually. Redundant properties are harmless and this check makes loading the file slower (probably not much, but still).

KoBeWi commented on editor/editor_settings.cpp:
```diff
@@ -1523,23 +1548,59 @@ Vector<String> EditorSettings::get_recent_dirs() const {
+			} else {
+				if (ResourceLoader::exists(E, "Script")) {
+					for (int i = 0; i < properties.size(); i++) {
+						if (!favorite_properties.has(properties[i])) {
```
```suggestion
						if (!favorite_properties[E].has(properties[i])) {
```
Although using HashSet would make this check unnecessary.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -37,6 +37,7 @@
 #include "editor/editor_feature_profile.h"
 #include "editor/editor_main_screen.h"
 #include "editor/editor_node.h"
+#include "editor/editor_paths.h"
```
What's this for?

YeldhamDev replied:
Blame my IDE for loving to automatically place includes everywhere.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -1267,6 +1295,15 @@ void EditorInspectorCategory::_notification(int p_what) {
 }
 
+void EditorInspectorCategory::override_menu(PopupMenu *p_menu) {
+	if (!p_menu) {
+		menu_override = nullptr;
+		return;
+	}
```
??

YeldhamDev replied:
I was probably thinking about adding more stuff here, didn't need to, but also forgot the clean up. Brain fart on my end.

KoBeWi commented:
The `menu_override` thing does not make sense. Instead of creating 2 menus for a category, you can either make it create proper menu (with constructor parameter) or add a getter for the menu and tweak it dynamically.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -3506,6 +3582,88 @@ void EditorInspector::update_tree() {
+
+		// Organize the favorited properties in their sections, to keep context and differentiate from others with the same name.
+		bool is_localized = property_name_style == EditorPropertyNameProcessor::STYLE_LOCALIZED;
+		for (HashMap<String, HashMap<String, LocalVector<EditorProperty *>>>::Iterator I = favorites_to_add.begin(); I;) {
```
Again with this strange iterator.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -4088,10 +4263,182 @@ void EditorInspector::_node_removed(Node *p_node) {
+				PropertyInfo &p = E->get();
+				if (p.usage & PROPERTY_USAGE_CATEGORY) {
+					path = p.hint_string;
+				} else if (p.usage & PROPERTY_USAGE_SCRIPT_VARIABLE) {
+					if (p.name == p_path) {
```
```suggestion
				} else if (p.usage & PROPERTY_USAGE_SCRIPT_VARIABLE && p.name == p_path) {
```

YeldhamDev commented:
@KoBeWi
> Favorite and Pin should be in the same separator group.

Should they though? They represent two completely different actions.

> I noticed Favorites sub-sections can't be folded. If it's not much problem, they should be foldable, unfolded by default.

My line of thought was that if you favorite a property, it's because you want it to be visible always. I'm not against making them foldable, but that would slightly increase code complexity, since they folded state would also need to be remembered like the rest. 

> Some weird bug

Fixed.

> The `menu_override` thing does not make sense. Instead of creating 2 menus for a category, you can either make it create proper menu (with constructor parameter) or add a getter for the menu and tweak it after creating.

Just a getter wouldn't work, as the code to popup the menu requires `doc_class_name` to not be empty. And about the constructor parameter, I didn't want to "spill out" the favorite stuff to a class that didn't need to know about it, so I added the override so the inspector would still be responsible for the implementation.

It's not a hill that I'm willing to die for, if you think the parameter is the best choice I will do it.

KoBeWi commented:
> Should they though? They represent two completely different actions.

They are similar enough. It's better than menu full of single-item separators that just looks ugly.

KoBeWi commented:
Patch that removes override menu:
[menu_no_override.txt](https://github.com/user-attachments/files/17398865/menu_no_override.txt)
The regular menu is created on demand. Makes the code simpler.

YeldhamDev commented:
Separator removed and iterators fixed.

YeldhamDev commented:
@KoBeWi I added your patch to the code.

KoBeWi reviewed (approved):
Looks good now.

After #97352 we can refer script favorites by UID instead of path, which would make the FileSystem Dock changes unnecessary.

Calinou reviewed (approved):
Tested locally, it works as expected.

![image](https://github.com/user-attachments/assets/d45076e1-05fd-4ab8-aa75-e9327c31efbb)

Note that this doesn't affect the remote inspector, which doesn't support favorite properties.

This makes me wonder if we should bring further attention to the Favorites section, e.g. by using a colored icon, or by adding a tinted background:

![image](https://github.com/user-attachments/assets/56d75a56-2a34-42ec-9266-388f5a1ba192)

![image](https://github.com/user-attachments/assets/d2bc14d2-4df7-41e0-b944-4bbb0b941eb3)

AThousandShips commented on editor/editor_settings.cpp:
```diff
@@ -1518,23 +1534,51 @@ Vector<String> EditorSettings::get_recent_dirs() const {
+		for (String &E : secs) {
+			PackedStringArray properties = PackedStringArray(cf->get_value(E, "properties"));
+			if (EditorNode::get_editor_data().is_type_recognized(E) || ResourceLoader::exists(E, "Script")) {
+				for (int i = 0; i < properties.size(); i++) {
+					if (!favorite_properties[E].has(properties[i])) {
+						favorite_properties[E].push_back(properties[i]);
```
```suggestion
				for (const String &property : properties) {
					if (!favorite_properties[E].has(property)) {
						favorite_properties[E].push_back(property);
```

AThousandShips commented on editor/editor_settings.cpp:
```diff
@@ -1518,23 +1534,51 @@ Vector<String> EditorSettings::get_recent_dirs() const {
+		}
+	}
+
+	/// Recent Directories
```
```suggestion
	// Recent Directories.
```

AThousandShips commented on editor/editor_inspector.cpp:
```diff
@@ -4088,10 +4256,169 @@ void EditorInspector::_node_removed(Node *p_node) {
+		String path;
+		HashMap<String, LocalVector<String>> props;
+
+		for (List<PropertyInfo>::Element *E = plist.front(); E; E = E->next()) {
+			PropertyInfo &p = E->get();
```
```suggestion
		for (PropertyInfo &p : plist) {
```

AThousandShips commented on editor/editor_inspector.cpp:
```diff
@@ -4088,10 +4256,169 @@ void EditorInspector::_node_removed(Node *p_node) {
+			scr->get_script_property_list(&plist);
+
+			String path;
+			for (List<PropertyInfo>::Element *E = plist.front(); E; E = E->next()) {
+				PropertyInfo &p = E->get();
```
```suggestion
			for (PropertyInfo &p : plist) {
```

AThousandShips commented on editor/editor_inspector.cpp:
```diff
@@ -4088,10 +4256,169 @@ void EditorInspector::_node_removed(Node *p_node) {
+		List<PropertyInfo> plist;
+		scr->get_script_property_list(&plist);
+
+		for (List<PropertyInfo>::Element *E = plist.front(); E; E = E->next()) {
+			PropertyInfo &p = E->get();
```
```suggestion
		for (PropertyInfo &p : plist) {
```

YeldhamDev commented on editor/editor_settings.cpp:
```diff
@@ -1518,23 +1534,57 @@ Vector<String> EditorSettings::get_recent_dirs() const {
 		recent_dirs_file = EditorPaths::get_singleton()->get_project_settings_dir().path_join("recent_dirs");
 	}
+
+	/// File Favorites
```
I dunno if it makes sense to add the period when it's capitalized like a title (which is why I used 3 slashes as well).

akien-mga reviewed (approved):
Looks good to me. I've been expecting this kind of feature for a long time!

Repiteo commented:
Thanks!

nuclear commented:
Could there be an option to have the favorited items show up at the top **and** where they used to be? I already see myself looking for some aspects because I forgot they're at the top.
